### ISSUE TITLE
Suggestion: Contract Programming

### ISSUE THREAD

**Issue by fdecampredon** [microsoft/TypeScript#198]
Contract Programming is a very nice way to prevent undetected failure, inspiring from [Dlang](http://dlang.org/contracts.html) here is a start of proposal : 

A new compiler flag `-contract` (or `-debug`) , without this flag the compiler would just ignore contract related code
### Function :

``` typescript
function myFunc(param: number): number 
  in {
    assert(param > 0); // the assertion library is not part of typescript but any thirdparty
  }
  out (result) {
    assert(result < 0);
  }
  body {
    return -param;
  }
```

would emit :

``` javascript
function myFunc(param) {
  (function () {
    assert(param > 0);
  })();

  var _$result = (function (param) {
     return -param;
  }).apply(this, arguments);

  (function (result) {
    assert(result < 0);
  }).call(this, _$result);

  return _$result;
}
```
### Class

respecting a D-like contract for class is challenging, especially rules about inheritance, and the outputed JS might become unreadable, here is a 'simplified' proposition without any inheritance logic. 

``` typescript

class MyClass {
  message: string = 'hello ';
  invariant {
   this.replaceMessage('test'); /// error invariant cannot call public member
   assert(typeof this.message === 'string');
  }

  constructor(name: string) 
    in {
      assert(typeof name === 'string');
    }
    //no out authorized for constructor
    body {
      this.message = this.message + name;
    }

  replaceMessage(message: string): string 
    in {
      assert(message);
    }
    out (result) {
      assert(result);
    }
    body {
      this.message = message;
      return this.message;
    }


  sayMessage() {
    alert(this.message);
  }
}

class ExtendedClass extends MyClass {
  invariant {
    assert(this.message.length > 0);
  }
}
```

would emit

``` javascript
var MyClass = (function () {
    function MyClass(name) {
        //call precondition
        (function (message) {
           assert(typeof name === 'string');
        ).apply(this, arguments);
        this.message = 'hello ';
        this.message = this.message + name;
        //call invariant
        if (this.constructor.__$invariant) {
           this.contructor.__$invariant.call(this);
        }
    }

    MyClass.prototype.replaceMessage = function (message) {
        //execute precondition
        (function (message) {
           assert(message);
        ).apply(this, arguments)
        //execute invariant
        if (this.constructor.__$invariant) {
           this.contructor.__$invariant.call(this);
        }
        //execute body
        var _$result = (function (message) {
           this.message = message;
           return this.message;
        }).apply(this, arguments);        
        //execute invariant   
        if (this.constructor.__$invariant) {
           this.contructor.__$invariant.call(this);
        }
        //execute out
        (function (result) {
           assert(result);
        }).call(this, _$result);
        return _$result;
    };

    MyClass.prototype.sayMessage = function () {
        if (this.constructor.__$invariant) {
           this.contructor.__$invariant.call(this);
        }
        //execute body
        var _$result = (function () {
           alert(this.message);
        }).apply(this, arguments);        
        //execute invariant        
        if (this.constructor.__$invariant) {
           this.contructor.__$invariant.call(this);
        }
        return _$result;
    };

    MyClass.__$invariant = function () {
      this.replaceMessage('test');
      assert(typeof this.message === 'string');
    } 
    return MyClass;
})();

var ExtendedClass = (function (_super) {
    __extends(ExtendedClass, _super);
    function ExtendedClass() {
        _super.apply(this, arguments);
    }

    ExtendedClass.__$invariant = function () {
        _super.__$invariant.call(this);
        assert(this.message.length > 0);
    } 
    return ExtendedClass;
})(MyClass);
```

**Comment by RichiCoder1** [comment]
:+1:
Came here out of curiosity after viewing the discussion for a similiar feature in C# over @ Roslyn.

A similar syntax as is being proposed over there might work:
```typescript
insert<T>(item: T, index: number)
    requires index >= 0 && index <= this.count
    ensures return >= 0 && return < this.count
{
    return this.insertCore(item, index);
}
```
which still comes down to the same:
```javascript
Array.prototype.insert = function(item, index) {
    if(!(index >= 0 && index <= this.count)) {
        throw "Failed requires.";
    }
    var __returnVal = this.insertCore(item, index);
    if(!(__returnVal >= 0 && __returnVal < this.count)) {
         throw "Failed ensures.";
    }
}
```
or
```javascript
Array.prototype.insert = function(item, index) {
    Assert.isTrue(index >= 0 && index <= this.count));
    var __returnVal = this.insertCore(item, index);
    Assert.isTrue(__returnVal >= 0 && __returnVal < this.count);
}
```

**Comment by nbransby** [comment]
Could decorators be used to support this? As suggested in the linked issue, something like:

@precondition("param > 0")
@postcondition("result < 0")
function myFunc(param: number): number {
    return -param;
}

Now I don't have the experience to say but it looks as if something like this work would without any languages changes? 

Perhaps you could even use proper expressions instead of string constants? Btw I'm just guessing here as I haven't taken the time to fully understand decorators but still interested in code contracts.

**Comment by eyalsk** [comment]
I love the idea, I've used CodeContracts in my C# projects for years but recently dropped it because the development is not  as active as it should be but really dislike the proposed syntax.

I really love the syntax proposed by @RichiCoder1 and iirc it's similar to SpecSharp.

**Comment by AlexGalays** [comment]
Does the typescript team even feels favorable to the very concept of contracts ?

I feel this might become more and more popular thanks to people now being exposed to things like clojure.spec. To me, it seems contracts/specs/etc value really (only) shine when used together with automatic/generative testing. Having to go to the proper screen, play with it till we reach the right state to trigger the proper function, to finally get our contract assert code trigger is not good and barely a step up from dynamic programming (React likes do to a lot of these runtime assertions)

Also, there might be some overlap with other tickets about dependent typing. Both have the same end goal: Express value constraints, with types or not.
While I assume runtime assertions would be way easier to implement, I feel that dependent types would be a much superior alternative in the absence of generative testing.

**Comment by eyalsk** [comment]
@AlexGalays 

> To me, it seems contracts/specs/etc value really (only) shine when used together with automatic/generative testing.

If they will introduce contracts into the language then they _might_ provide static analysis but in case they won't then anyone else can pick it up and do it but first it needs to be formalized in the language.

**Comment by omouse** [comment]
@AlexGalays just to add more information, contract-based programming has existed in Eiffel for a very long time and that's a language that compiles to C. They [provide different compilation options to check assertions and contracts at run-time](https://www.eiffel.org/doc/eiffel/ET%3A%20Design%20by%20Contract%20(tm),%20Assertions%20and%20Exceptions#Run-time_assertion_monitoring). You can use contracts to generate tests and documentation, but run-time assertions are something I would highly recommend. You want to fail fast and visibly (even in user-facing applications).

**Comment by electricessence** [comment]
In a way, TypeScript is doing this.  StrictNullChecks for example will let you know upstream if you are going to cause a null reference. :)

**Comment by isiahmeadows** [comment]
@electricessence That's closer to the distinction between `Type` and `Maybe<Type>` (where the latter is defined as `type Maybe<T> = T | void`). Constraints are different. There's a much greater difference between a sugared `Type? = Maybe<Type>` and `param: Type where param > 10`. Additionally, TypeScript retains an extraordinarily high bar on what features merit an actual emit. (Type-safe enums and namespaces hit that bar when they were first introduced, but integer casts haven't hit even close yet.)

As for where I stand, I'd prefer something that allows constraints to be verified at compile-time *without* an emit, but implementing it would be far from trivial best case scenario, and the compiler, from what I've gathered, is a massive hack in its current state.

**Comment by electricessence** [comment]
@isiahmeadows I'm with you.  I'm just happy to see it doing something similar at the moment.
Would love to see full contract capability.

**Comment by Deilan** [comment]
Looking forward to it!

**Comment by josefbetancourt** [comment]
Just for reference.  Contractual uses a compiler to generate the contracts.  If compiler not used, the extra code remain as unused objects. (maybe those are eventually stripped by normal JS runtime or ts).  See  http://codemix.github.io/contractual/index.html#usage

**Comment by eyalsk** [comment]
@josefbetancourt Most DBC implementations exist as a compile-time feature whether the ability to express the contracts is baked into the language and whether an analyzer is built-in to the compiler but it's definitely not a run-time feature so I'm not sure what you're on about.

**Comment by pawsaw** [comment]
@fdecampredon It's a long time since this feature request... do you know the current state of "design by contract" in TypeScript... maybe something as a third party library you could suggest?

**Comment by petermetz** [comment]
> 
> 
> @fdecampredon It's a long time since this feature request... do you know the current state of "design by contract" in TypeScript... maybe something as a third party library you could suggest?

I'd love to see a library that mimics the `Preconditions` utility class from the Java library called Guava. Would need to be tailored a little bit to the language specifics, but for the most part it would be great just as is IMHO.
Been considering writing something like that myself for a long time, but time is always an issue of course.

Big thumbs up for this thread in general. 👍

**Comment by Mouvedia** [comment]
@petermetz https://github.com/codemix/contractual

**Comment by RyanCavanaugh** [comment]
This is very much outside our modern design goals.

**Comment by kasir-barati** [comment]
@RyanCavanaugh Really?

> This is very much outside our modern design goals.

When I start reading this issue I had hope that eventually I will see that this feature request is rejected or said that it is planned to be done in the future. BTW I guess TS needs this feature in a declarative manner. IMO a good syntax belongs to clojure:

```clj
(def deposit)

(defn account-transactions [account-id]
   #{"1000" "1001" "1002"} )

(defn create-transaction [account, type, amount]
  (println account type amount)
  "1001"
  )

(defn account-open? [account-id]
  true)

;
(defn accept-deposit [account-id amount]
   { :pre [  (> amount 0.00) 
             (account-open? account-id) ]
     :post [ (contains? (account-transactions account-id) %) ] }
   "Accept a deposit and return the new transaction id"
   ;; Some other processing goes here...
   ;; Return the newly created transaction:
   (create-transaction account-id :deposit amount))
;

(accept-deposit "1001" 422.00) ; Ok
(accept-deposit "1002" -422.00) ; should blow up
```

Please share some ref if it is planned to be added in the language. Or tell why it is not fulfill able for now?

**Comment by RyanCavanaugh** [comment]
See non-goal 6 https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

> Provide additional runtime functionality
